### PR TITLE
[WIP] Pure swift MySQL

### PR DESCRIPTION
**Pull request by Joannis**
Done:

- [x] Type-safe results
- [x] String Queries
- [x] Authentication
- [x] Async
- [x] Basic Connection pooling
- [x] Streams
- [x] Futures

TODO:

- [ ] Tests
- [ ] SSL
- [x] Raw rows

Bugs:

- [x] empty results are never complete

### CODE REVIEW

File: Documents/Manifestos/MYSQL.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+## Introduction
+
+MySQL is a popular for web services. It's widely adopted and even familiar under many NoSQL database users.
+
+Thid document serves as a lead for discussing, implementing and accepting code for the `MySQL` package.
+
+## Project goals
+
+### Asynchronous
+
```
Comment (tanner0101) [inline Documents/Manifestos/MYSQL.md]: not just async, but also non-blocking

File: Sources/MySQL/Internals/Internals.swift [hunk 2]
```diff
@@ -15,10 +16,40 @@ enum Response {
...
     }
     
-    struct Error : Swift.Error {
+    struct Error : Swift.Error, Traceable, Debuggable {
+        var reason: String {
+            return message
+        }
+        
+        var identifier: String {
+            return "mysql:error:\(message)"
```
Comment (tanner0101) [inline Sources/MySQL/Internals/Internals.swift]: `MySQL.Error` will already be prepended to the identifier. There should be a 1-20ish char identifier here (like an enum case key)

File: Sources/MySQL/Query/Schema.swift [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    public var schema = [Column]()
+    
+    public init(named name: String, temporary: Bool = false) {
+        self.name = name
+        self.temporary = temporary
+    }
+}
+
+extension ConnectionPool {
+    public func createTable(_ table: Table) throws -> Future<Void> {
```
Comment (tanner0101) [inline Sources/MySQL/Query/Schema.swift]: This function makes more sense accepting a `Schema`. When I see `Table`, I think of a table that already exists. I think `Table` would be better as a protocol that you conform your models to.

File: Sources/MySQL/Results/ResultsStream+Model.swift [hunk 4]
```diff
@@ -11,6 +11,7 @@ public class ModelStream<D: Decodable> : ResultsStream {
         self.mysql41 = mysql41
     }
     
+    /// Internal API: Do not touch
     public var onClose: (() -> ())?
```
Comment (tanner0101) [inline Sources/MySQL/Results/ResultsStream+Model.swift]: Any way we can make this `internal`?

File: Sources/SQLQuery/Administration.swift [hunk 5]
```diff
@@ -0,0 +1,10 @@
+public enum AdministrationQuery {
```
Comment (tanner0101) [inline Sources/SQLQuery/Administration.swift]: This should probably be on connection pool, or the methods on connection pool should also be on an enum

File: Sources/MySQL/Connection/ConnectionPool.swift [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        self.port = port
+        self.user = user
+        self.password = password
+        self.database = database
+    }
+    
+    typealias Complete = (()->())
+    
+    /// Retains a connection (or creates a new one) to execute the handler with
+    internal func retain<T>(_ handler: @escaping ((Connection, @escaping ((T) -> ()), @escaping Stream.ErrorHandler) -> Void)) throws -> Future<T> {
```
Comment (tanner0101) [inline Sources/MySQL/Connection/ConnectionPool.swift]: typealiases could make this easier to read

File: Sources/MySQL/Connection/ConnectionPool.swift [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+                pair.reserved = false
+                promise.complete(result)
+            }) { error in
+                pair.reserved = false
+                promise.fail(error)
+            }
+        }
+        
+        return promise.future
+    }
```
Comment (tanner0101) [inline Sources/MySQL/Connection/ConnectionPool.swift]: this function is a bit long and hard to parse. can we separate into smaller bits?

File: Sources/MySQL/Errors/Error.swift [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        return self.problem.rawValue
+    }
+    
+    /// Creates a new problem
+    init(_ problem: Problem,
+         file: String = #file,
+         function: String = #function,
+         line: UInt = #line,
+         column: UInt = #column
+        ) {
```
Comment (tanner0101) [inline Sources/MySQL/Errors/Error.swift]: spacing

File: Sources/MySQL/Internals/Handshake.swift [hunk 9]
```diff
@@ -129,20 +96,23 @@ import Crypto
...
             self.handshake = handshake
             
             try self.sendHandshake()
         } catch {
+            self.authenticated.fail(error)
             self.socket.close()
         }
     }
     
+    /// Send the handshaek to the client
```
Comment (tanner0101) [inline Sources/MySQL/Internals/Handshake.swift]: typo

File: Sources/MySQL/Connection/ConnectionPool.swift (hunk 6 above)
Comment (Joannis) [reply Sources/MySQL/Connection/ConnectionPool.swift]: I agree, but that doesn't work nicely with generics

File: Sources/MySQL/Connection/ConnectionPool.swift (hunk 7 above)
Comment (Joannis) [reply Sources/MySQL/Connection/ConnectionPool.swift]: It's really tough to split up. Then I tried to split it up I ended up with much less readable code.
